### ISSUE TITLE
spurious std lib test failures with -OReleaseSmall. branching on undefined value in Sema

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#12933]
Zig Version: 0e4b04672c000220ef47894086100610a9d6d9c3

Example CI failure:
```
error: test...
error: The following command terminated unexpectedly:
/workspace/_release/staging/bin/zig test /workspace/lib/std/std.zig --test-name-prefix std-native-ReleaseSmall-bare-single-default  -OReleaseSmall --cache-dir /workspace/zig-cache --global-cache-dir /root/.cache/zig --name test -fsingle-threaded -I /workspace/test -L /deps/local/lib -isystem /deps/local/include --zig-lib-dir /workspace/lib --enable-cache 
```

I ran valgrind on it and got some warnings:

```
[nix-shell:~/Downloads/zig/build]$ valgrind stage3/bin/zig test ../lib/std/std.zig -OReleaseSmall
==2557576== Memcheck, a memory error detector
==2557576== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==2557576== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==2557576== Command: stage3/bin/zig test ../lib/std/std.zig -OReleaseSmall
==2557576== 
Semantic Analysis [92] testing... ==2557576== Conditional jump or move depends on uninitialised value(s)
==2557576==    at 0x9E315A3: type.Type.ptr (type.zig:6414)
==2557576==    by 0x9E32C3F: Sema.analyzeDeclRef (Sema.zig:26351)
==2557576==    by 0xA0339E4: Sema.analyzeDeclVal (Sema.zig:26264)
==2557576==    by 0x9DFCFEC: Sema.zirThis (Sema.zig:13912)
==2557576==    by 0x9BD8246: Sema.analyzeBodyInner (Sema.zig:929)
==2557576==    by 0x9A12718: Sema.analyzeBodyBreak (Sema.zig:628)
==2557576==    by 0xA025D85: Sema.resolveBody (Sema.zig:593)
==2557576==    by 0x9D75763: Sema.zirCall (Sema.zig:5671)
==2557576==    by 0x9BCE590: Sema.analyzeBodyInner (Sema.zig:727)
==2557576==    by 0x9E2B031: Sema.zirCondbr (Sema.zig:15279)
==2557576==    by 0x9BDEF29: Sema.analyzeBodyInner (Sema.zig:1347)
==2557576==    by 0xA05FE12: Sema.resolveBlockBody (Sema.zig:4939)
==2557576== 
==2557576== Conditional jump or move depends on uninitialised value(s)
==2557576==    at 0x9E315EB: type.Type.ptr (type.zig:6414)
==2557576==    by 0x9E32C3F: Sema.analyzeDeclRef (Sema.zig:26351)
==2557576==    by 0xA0339E4: Sema.analyzeDeclVal (Sema.zig:26264)
==2557576==    by 0x9DFCFEC: Sema.zirThis (Sema.zig:13912)
==2557576==    by 0x9BD8246: Sema.analyzeBodyInner (Sema.zig:929)
==2557576==    by 0x9A12718: Sema.analyzeBodyBreak (Sema.zig:628)
==2557576==    by 0xA025D85: Sema.resolveBody (Sema.zig:593)
==2557576==    by 0x9D75763: Sema.zirCall (Sema.zig:5671)
==2557576==    by 0x9BCE590: Sema.analyzeBodyInner (Sema.zig:727)
==2557576==    by 0x9E2B031: Sema.zirCondbr (Sema.zig:15279)
==2557576==    by 0x9BDEF29: Sema.analyzeBodyInner (Sema.zig:1347)
==2557576==    by 0xA05FE12: Sema.resolveBlockBody (Sema.zig:4939)
==2557576== 
==2557576== Conditional jump or move depends on uninitialised value(s)
==2557576==    at 0x9E3171E: type.Type.ptr (type.zig:6429)
==2557576==    by 0x9E32C3F: Sema.analyzeDeclRef (Sema.zig:26351)
==2557576==    by 0xA0339E4: Sema.analyzeDeclVal (Sema.zig:26264)
==2557576==    by 0x9DFCFEC: Sema.zirThis (Sema.zig:13912)
==2557576==    by 0x9BD8246: Sema.analyzeBodyInner (Sema.zig:929)
==2557576==    by 0x9A12718: Sema.analyzeBodyBreak (Sema.zig:628)
==2557576==    by 0xA025D85: Sema.resolveBody (Sema.zig:593)
==2557576==    by 0x9D75763: Sema.zirCall (Sema.zig:5671)
==2557576==    by 0x9BCE590: Sema.analyzeBodyInner (Sema.zig:727)
==2557576==    by 0x9E2B031: Sema.zirCondbr (Sema.zig:15279)
==2557576==    by 0x9BDEF29: Sema.analyzeBodyInner (Sema.zig:1347)
==2557576==    by 0xA05FE12: Sema.resolveBlockBody (Sema.zig:4939)
==2557576== 
```

It suggests a problem with this condition:

```
        if (d.@"align" != 0 and d.@"align" == d.pointee_type.abiAlignment(target)) {
```

and this one:

```
        if (d.@"align" == 0 and d.@"addrspace" == .generic and
```

Which suggests that at the bottom of `analyzeDeclRef` it is copying alignment value from a Decl which is not yet initialized.

**Comment by andrewrk** [comment]
Furthermore, when I built the std lib tests with a release build of zig running in valgrind (same command above, but release stage3 instead of debug stage3), I observed two test failures:

```
Test [586/2207] test.math.isFinite... FAIL (TestUnexpectedResult)
Test [590/2207] test.math.isNormal... FAIL (TestUnexpectedResult)
```

**Comment by jacobly0** [comment]
Reduction for branch on undefined value:
```zig
fn ref(comptime _: type) void {}
test {
    ref(@This());
}
```

**Comment by andrewrk** [comment]
Fix landed in 54eb0f2daa2cf5237da05e3cfad5800b9ccf424e.
